### PR TITLE
Add MURMULATOR platform & Fix I/O

### PR DESCRIPTION
**Pull request by javavi**
Added MURMULATOR platform (HDMI & VGA variant)
https://t.me/ZX_MURMULATOR
nMOS Zilog Z80 processor emulation enabled, as in the original Sinclair ZX-Spectrum
Fix W/R IO port  (games that didn't work started working)
Fixed envelope shapes AY
Add AY STEREO mode & Speeker Volume control
. . .

### CODE REVIEW

**Comment by fruit-bat** [comment]
This looks great. I need to check it still works with the other targets.  Unfortunately,  I am quite unwell at the moment, so bear with me. Just wanted you to know that the pull request is appreciated.

File: src/ZxSpectrum.cpp [hunk 1]
```diff
@@ -16,13 +16,17 @@ ZxSpectrum::ZxSpectrum(
...
   _borderColour(7),
+  _port254(0),
   _ear(false),
   _earInvert(0),
   _earDc(0),
   _buzzer(0)
 {
   z80Power(true);
   _Z80.context = this;
+#ifdef MURMULATOR
```
Comment (fruit-bat) [inline src/ZxSpectrum.cpp:27]: Don't bother with the #define if you believe this is more correct. Just set it.

File: src/ZxSpectrum.h [hunk 2]
```diff
@@ -81,7 +81,36 @@ class ZxSpectrum {
     if (address < 0x4000) return;    
     *(memaddr(address)) = value;
   }
-  
+
+#ifdef MURMULATOR
```
Comment (fruit-bat) [inline src/ZxSpectrum.h:85]: Always use this version if you believe it is more correct.

File: src/ZxSpectrum.h [hunk 3]
```diff
@@ -107,7 +136,42 @@ class ZxSpectrum {
       return 0xff;
     }
   }
-  
+  #endif
+
+#ifdef MURMULATOR
```
Comment (fruit-bat) [inline src/ZxSpectrum.h:141]: Always use this version if you believe it is more correct.

File: src/ZxSpectrumAy.h [hunk 4]
```diff
@@ -36,9 +36,15 @@ static const uint8_t Envelopes[16][32] =
 };
 
 // See https://github.com/retrofw/speccy/blob/master/devices/sound/ay.cpp
+#ifdef MURMULATOR
```
Comment (fruit-bat) [inline src/ZxSpectrumAy.h:39]: Always use this version if you believe it is more correct.

File: src/ZxSpectrumMenu.cpp [hunk 5]
```diff
@@ -340,6 +340,7 @@ ZxSpectrumMenu::ZxSpectrumMenu(
   };
 
   onPaint([](PicoPen *pen) {
+     pen->printAt(0, 0, false, "ZX Spectrum 48K/128K by fruit-bat on MURMULATOR");
```
Comment (fruit-bat) [inline src/ZxSpectrumMenu.cpp:343]: Add missing #ifdef

File: src/murmulator/ps2kbd_mrmltr.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
+// You may use, distribute and modify this code under the
```
Comment (fruit-bat) [inline src/murmulator/ps2kbd_mrmltr.h:1]: This is similar to code in pico-emu-utils ps2kbd.cpp, ps2kbd.h & ps2kbd.pio
https://github.com/fruit-bat/pico-emu-utils
Please reuse if possible.

File: src/murmulator/ps2kbd_mrmltr.pio [hunk 7]
```diff
@@ -0,0 +1,23 @@
+; Copyright (C) 1883 Thomas Edison - All Rights Reserved
```
Comment (fruit-bat) [inline src/murmulator/ps2kbd_mrmltr.pio:1]: This is similar to code in pico-emu-utils ps2kbd.cpp, ps2kbd.h & ps2kbd.pio
https://github.com/fruit-bat/pico-emu-utils
Please reuse if possible.

**Review by fruit-bat — COMMENTED** [review]
I've added some review comments. Hopefully, you are happy to action them. Let me know if  you are having any problems.

**Comment by javavi** [comment]
Thank you for your attention to my request!
I am building a project on Pico SDC Toolchai on Windows 10.

- In my environment, the new CMake version 3.25.2 is installed, which
displays 8 warnings even before compilation.
cmake_policy(SET CMP0115 OLD)
hides these warnings.

- Changes made to ZxSpectrum.h fixed input and output ports, which made it
possible to run a paradise of games.
- Changes made to ZxSpectrumAy.h made it possible to achieve the correct
sound.

- Changes in ZxSpectrumMenu.cpp are more for the amusement of the local
audience :)

- A problem with the PS/2 keyboard driver resulted in the need to replace
the library and PIO file.
   In the MURMULATORA scheme, the DATA bits are not in the same sequence as
the library processes them.

There are a number of issues that I am currently working on and need your
advice.
- 60Hz interrupt rate is not native to ZX Spectrum which has a 50Hz
interrupt rate
   Because of this, the music is distorted, plays faster, a number of games
crash.
I've experimented with different approaches to convert interrupts to 50Hz.
Music, gameplay restored to normal, but interrupts are out of sync with
frame rate.
Created a border buffer to display the border correctly but the border
effects are jittery, maybe
there is no clear stability in the execution time of processor command
cycles.
A more correct approach for negotiation is a frame buffer into which the
input process writes to
line emulation frequency of 50Hz, and the output process at VGA 60Hz.
This approach is used by the authors of emulators for MURMULATOR and gives
a good result.

пн, 29 мая 2023 г. в 14:46, fruit-bat ***@***.***>:

**Comment by javavi** [comment]
I will follow your instructions and recommendations, make changes to the
code, and soon execute another Pull Request!
Thank you for your attention to my requests!

вт, 30 мая 2023 г. в 21:55, Vadim Jatlov ***@***.***>:

**Comment by fruit-bat** [comment]
> 

You can still make changes to this pull request. No need to make a new one 🙂

**Comment by fruit-bat** [comment]
I've created a new local branch based on your changes here:

https://github.com/fruit-bat/pico-zxspectrum/tree/feature/murmulator

I have also made a new pull request back to main from this branch.

On the branch I've made *some* of the changes I suggested in the pull request review. I have added a couple of executables. If you can confirm if they work correctly it would be appreciated.

If you wish to make changes I suggest you get the latest version of my code and switch to the murmulator branch, then pull request changes to my branch. That way we can both work on the code without committing to main :-) At some point I will close this pull request with out merging, as your code will get to main via the branch.

**Comment by fruit-bat** [comment]
I hope you don't mind but I have raised an issue about the 60Hz refresh rate on your behalf. I was worried the conversation was going to be lost in this pull request.

**Comment by javavi** [comment]
Good health and spirit!
In an attempt to make 50Hz interrupts normal for the ZX Spectrum, I tied to
the line counter from the process (core1_main). Apparently the
get_vga_line() function is only executed for rendered lines and does not
execute vertical sync blanking lines. From here, apparently, a jitter of
the interrupt period of 24 lines is obtained.
Maybe you have some ideas where to get the counter of all lines of the
video output process?

ср, 7 июн. 2023 г. в 09:21, fruit-bat ***@***.***>:

**Pull request merged**